include/revng/Support/GraphAlgorithms.h

```diff
@@ -518,26 +541,844 @@ nodesBetweenNewImpl(GraphT Source,
+  for (auto RegionIt1 = R.begin(); RegionIt1 != R.end(); RegionIt1++) {
+    for (auto RegionIt2 = std::next(RegionIt1); RegionIt2 != R.end();
+         RegionIt2++) {
+      bool Intersects = intersect(*RegionIt1, *RegionIt2);
+      bool IsIncluded = subset(*RegionIt1, *RegionIt2);
+      bool IsIncludedReverse = subset(*RegionIt2, *RegionIt1);
+      bool AreEquivalent = equal(*RegionIt1, *RegionIt2);
+      if (Intersects
+          and (((!IsIncluded) and (!IsIncludedReverse)) or AreEquivalent)) {
+        (*RegionIt1).insert((*RegionIt2).begin(), (*RegionIt2).end());
+        R.erase(RegionIt2);
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+template<class NodeT>
+void simplifyRegions(llvm::SmallVector<SmallSetVector<NodeT>> &Rs) {
+  bool Changes = true;
+  while (Changes) {
+    Changes = simplifyRegionsStep(Rs);
+  }
+}
+
+// Reorder the vector containing the regions so that they are in increasing size
+// order.
+template<class NodeT>
+void sortRegions(llvm::SmallVector<SmallSetVector<NodeT>> &Rs) {
+  std::sort(Rs.begin(),
+            Rs.end(),
+            [](SmallSetVector<NodeT> &First, SmallSetVector<NodeT> &Second) {
+              return First.size() < Second.size();
+            });
+}
+
+template<class GraphT,
+         class GT = llvm::GraphTraits<GraphT>,
+         typename NodeRef = typename GT::NodeRef>
+SmallDenseMap<NodeRef, size_t> computeDistanceFromEntry(GraphT Source) {
+  SmallDenseMap<NodeRef, size_t> ShortestPathFromEntry;
+
+  using SetType = llvm::bf_iterator_default_set<NodeRef>;
+  using bf_iterator = llvm::bf_iterator<GraphT, SetType, GT>;
+  auto BFSIt = bf_iterator::begin(Source);
+  auto BFSEnd = bf_iterator::end(Source);
+
+  for (; BFSIt != BFSEnd; ++BFSIt) {
+    NodeRef Block = *BFSIt;
+    size_t Depth = BFSIt.getLevel();
+
+    // Obtain the insertion iterator for the `Depth` block element.
+    auto ShortestIt = ShortestPathFromEntry.insert({ Block, Depth });
+
+    // If we already had in the map an entry for the current block, we need to
+    // assert that the previously found value for the `Depth` is less or equal
+    // of the `Depth` we are inserting.
+    if (ShortestIt.second == false) {
+      revng_assert(ShortestIt.first->second <= Depth);
+    }
+  }
+
+  return ShortestPathFromEntry;
+}
+
+template<class NodeT>
+bool setContains(SmallSetVector<NodeT> &Set, NodeT &Element) {
+  return Set.contains(Element);
+}
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+auto child_range(GraphT Block) {
+  return llvm::make_range(GT::child_begin(Block), GT::child_end(Block));
+}
+
+template<class GraphT>
+auto successor_range(GraphT Block) {
+  return child_range(Block);
+}
+
+template<class GraphT>
+size_t successor_range_size(GraphT Block) {
+  auto Range = successor_range(Block);
+  return std::distance(Range.begin(), Range.end());
+}
+
+template<class GraphT>
+GraphT get_successor(GraphT Block, ptrdiff_t I) {
```

Comment:
1. I changed this in bd8b633422434d6d23731090767b287409f236f1.
2. The point of having the `getBlockSuccessor` laid out like this is mainly in order to be compatible with all the template based infrastructure. This seems to me the only way of requiring only `GraphTraits` implemented on `GraphT` and be able to extract the n-th successors. Are you suggesting to always materialize the list of successors on the call-site side of the thing?

Reply:
2. I think `get_successor` is a "dangerous" API to offer, since it might lead you into writing something like
   ```c++
   for (unsigned I = 0; I < ...; ++I)
     doSomething(getSuccessor(Graph, I));
   ```
   So, yes, I'm suggesting not to offer this API and let the user do this by hand if he really really needs to.

Reply:
Ah ok, now your point is clear. I removed the function in 0ff3e6296c9aecc1bd81bf7c2532dfcbabd8fdfd, and I materialize this when needed.